feat: add input function by typing to timepicker

Genne23v opened the pull request:
## Proposed changes
<!-- Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue. -->
Following up [issue-118 Time Picker allow for typing input](https://github.com/materializecss/materialize/issues/118), I have added time input by typing functionality. The user can edit the time and minute on the digital clock directly while original functionality remains same. I have updated `/js/timepicker.js`. But I tested with `docs/js/materialize`. Please let me know if I picked up the wrong file. Once my direction is accepted, I will also add tests and update the documentation as well. And I would appreciate any feedback or other direction. I will quickly follow up. 

## Screenshots (if appropriate) or codepen:
<!-- Add supplemental screenshots or code examples. Look for a codepen template in our **[CONTRIBUTING document](https://github.com/materializecss/materialize/blob/master/CONTRIBUTING.md)**. -->
![Screen Shot 2022-10-09 at 7 47 36 PM](https://user-images.githubusercontent.com/59734889/194785010-c28dc8cc-5a3f-48f7-8cc0-4b5878b4eb22.png)


## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to change).

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have read the **[CONTRIBUTING document](https://github.com/materializecss/materialize/blob/main/CONTRIBUTING.md)**.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

LoganTann commented on js/timepicker.js:
```diff
@@ -485,6 +495,65 @@
     }
 
+    _inputFromTextField() {
+      var isHours = this.currentView === 'hours';
+      var unit = Math.PI / (isHours ? 6 : 30);
+      var radius, radian;
```
prefer the use of `const` (and `let` for the variables that are reassigned) instead of `var`

Genne23v replied:
Totally agree. I will update accordingly

LoganTann commented on js/timepicker.js:
```diff
@@ -485,6 +495,65 @@
+      if (isHours) {
+        var value = this['inputHours'].value;
+
+        if (!(value > 0 && value < 13)) {
+          this['inputHours'].value = new Date().getHours();
+        } else {
+          radian = value * unit;
+
```
As there is a  
`if (not something) {doThis} else {doThat}`,

wouldn't it be easier to just do  
`if (something) {doThat} else {doThis}` ?

Genne23v replied:
Also agreed. Will be addressed.

LoganTann commented on js/timepicker.js:
```diff
@@ -485,6 +495,65 @@
+        } else {
+          radian = value * unit;
+
+          var cx1 = Math.sin(radian) * (radius - this.options.tickRadius),
+            cy1 = -Math.cos(radian) * (radius - this.options.tickRadius),
+            cx2 = Math.sin(radian) * radius,
+            cy2 = -Math.cos(radian) * radius;
+
+          this.hand.setAttribute('x2', cx1);
+          this.hand.setAttribute('y2', cy1);
+          this.bg.setAttribute('cx', cx2);
+          this.bg.setAttribute('cy', cy2);
```
Can you try to move as much redundant code as possible into separate methods ?

Genne23v replied:
Will be addressed

LoganTann commented on js/timepicker.js:
```diff
@@ -485,6 +495,65 @@
+          this.spanMinutes.focus();
+        }
+      } else {
+        var value = this['inputMinutes'].value;
+
+        if (!(value >= 0 && value < 60)) {
+          var minutes = new Date().getMinutes();
+          this['inputMinutes'].value = Timepicker._addLeadingZero(minutes);
+        } else {
```
Same comments as previous ones : try to use const, invert the if to simplify the condition and if possible, refactor the code to move large blocks into separate methods.

Genne23v replied:
Agree same as above

LoganTann commented on sass/components/_timepicker.scss:
```diff
@@ -11,7 +11,7 @@
 
 .text-primary {
-	color: rgba(255, 255, 255, 1);
+	color: rgba(255, 255, 255, 1) !important;
```
Do you really need the `!important` specifier ?

As it should be mostly used as a tool of last resort (since it breaks the principle of *cascading* stylesheet), try to use a more specific selector.

Genne23v replied:
`!important` is not necessary here. I will remove it.

LoganTann commented on sass/components/_timepicker.scss:
```diff
@@ -24,13 +24,18 @@
 
 .timepicker-text-container {
-	font-size: 4rem;
+  width: 200px;
+  font-size: 4rem;
 	font-weight: bold;
 	text-align: center;
 	color: rgba(255, 255, 255, 0.6);
   font-weight: 400;
 	position: relative;
   user-select: none;
+  input[type=text]{
+    color: rgba(255, 255, 255, 0.6);
+    direction: rtl;
+  }
```
please fix indentation of the whole block

Genne23v replied:
Will be fixed

LoganTann replied:
Indentation of lines 28-30 is inconsistent.

Genne23v replied:
My editor shows it's aligned somehow, but I will fix accordingly.

LoganTann commented on sass/components/_timepicker.scss:
```diff
@@ -39,12 +44,20 @@
 
-.timepicker-span-hours {
-  margin-right: 3px;
+.timepicker-input-hours {
+  height: 4rem !important;
+  width: 30% !important;
+  border-bottom: 0px !important;
+  font-size: 4rem !important;
+  margin-right: 3px !important;
 }
 
```
Same as previous comments. `!important` should be used with partiality.

LoganTann commented on sass/components/_timepicker.scss:
```diff
@@ -39,12 +44,20 @@
 
-.timepicker-span-minutes {
-  margin-left: 3px;
+input.timepicker-input-minutes {  //change class name
```
Did you forget to remove the `// change class name` comment ?

Genne23v replied:
Sorry for leaving unnecessary comment. It will be removed.

Genne23v commented:
Please review my updated code along with my demo on codepen. Thanks! 
https://codepen.io/genne23v/pen/oNdmgvM?editors=1111

LoganTann commented on js/timepicker.js:
```diff
@@ -496,64 +496,59 @@
 
       if (isHours) {
-        var value = this['inputHours'].value;
+        let value = this['inputHours'].value;
```
Is value reassigned ? if not, prefer the use of const. Same comment for line 516

Genne23v replied:
Yes, it should be const

LoganTann commented on js/timepicker.js:
```diff
@@ -116,14 +116,16 @@
       this._handleClockClickStartBound = this._handleClockClickStart.bind(this);
       this._handleDocumentClickMoveBound = this._handleDocumentClickMove.bind(this);
       this._handleDocumentClickEndBound = this._handleDocumentClickEnd.bind(this);
+      this._inputFromTextFieldBound = this._handleTimeInputEnterKey.bind(this);
 
```
User experience advice : the clock should be updated when typing (keyup) or on input focus lost, instead of pressing the enter key.

I will not block the merge due to this comment, as this is not the scope of the issue yet.

Genne23v replied:
I think I get what you mean. I will just update `keydown` to `keyup`.

LoganTann commented:
Code is approved ! Thanks again.

However, the code quality pipeline will fail due to a wrong commit title format.

> You either need to rename them or create a new pull request with a fresh commit in order for the pipeline to pass, unfortunately.
> 
> To generate a commit message that respect the convention, you can use https://commitlint.io/

Genne23v commented:
> 

I have updated PR title. I'm not sure if this is going to pass the pipeline. Please let me know. Thanks

LoganTann commented:
> > 
> 
> I have updated PR title. I'm not sure if this is going to pass the pipeline. Please let me know. Thanks

No, by "title" I meant the commit titles 👇 

![image](https://user-images.githubusercontent.com/28659185/195928718-e0f1ff3c-9992-4ba6-973b-1775cef99352.png)

A possible way to fix this is to : 
1. make a local copy of your code somewhere
2. return to the main branch and create a new branch
3. paste the code
4. commit the whole using the correct commit title
5. Open a new Pull Request

Or by doing a git rebase (see https://stackoverflow.com/a/41987851)

Anyway, make sure to make a save before applying the steps I mentioned before, as the steps involves rewriting your changes.

Genne23v commented:
> > > 
> > 
> > 
> > I have updated PR title. I'm not sure if this is going to pass the pipeline. Please let me know. Thanks
> 
> No, by "title" I meant the commit titles 👇
> 
> ![image](https://user-images.githubusercontent.com/28659185/195928718-e0f1ff3c-9992-4ba6-973b-1775cef99352.png)
> 
> A possible way to fix this is to :
> 
> 1. make a local copy of your code somewhere
> 2. return to the main branch and create a new branch
> 3. paste the code
> 4. commit the whole using the correct commit title
> 5. Open a new Pull Request
> 
> Or by doing a git rebase (see https://stackoverflow.com/a/41987851)
> 
> Anyway, make sure to make a save before applying the steps I mentioned before, as the steps involves rewriting your changes.

I removed the failing branch. Please let me know if this works. Sorry for any inconvenience.

LoganTann commented:
Sorry, there are still two commits that are not valid...

![image](https://user-images.githubusercontent.com/28659185/196003187-3c6ee1d1-e1c9-4541-95a3-d1870e977750.png)

Genne23v commented:
I removed other commits too. Hope this passes. Thanks

LoganTann reviewed (approved):
all good ! Last question, are you doing this MR as part of the hacktoberfest ?

Genne23v commented:
> 
Thanks for the approval. Yes, it's a part of hacktoberfest.

LoganTann commented:
Thank you a lot for taking your time on this valuable contribution, this project needed it.

It was a bit time-wasting with this story of commit title, but we've made it finally!

Genne23v commented:
> 

It's my pleasure since it's my first open source contribution. I think I will remember this PR for long time. It's really nice framework and I hope I will contribute more to this. Thank you!

Smankusors commented:
btw, I invited you as the member of materializecss. Welcome! 🎉

Genne23v commented:
> 

Thank you for the invite. I'm looking forward to more contributions!

AurelienTisserand commented:
Hello, I just migrate to 1.2.0 and I tested this functionnality but it seems it doesn't work on my side. I can edit time value manually but it doesn't update clock and when I save hour is not save as I choose. As you can see in the following screenshot I put '30' manualy via my keyboard and it doesn't do anything even if I left focus of the input. Did I misunderstand this new feature ?

![image](https://user-images.githubusercontent.com/36203816/203912891-b6f031c2-0d7b-4d77-be04-e6f8e9d805e9.png)

Smankusors commented:
@AurelienTisserand you need to press enter on the keyboard to confirm it

AurelienTisserand commented:
> @AurelienTisserand you need to press enter on the keyboard to confirm it

Thank you for you answer @Smankusors but on my side pressing enter is submitting my form without updating the date. Did I miss something ?

Smankusors commented:
> Thank you for you answer @Smankusors but on my side pressing enter is submitting my form without updating the time. Did I miss something ?

hmm... can you create a small reproduceable codepen/jsfiddle for this? 😅
